### DISCUSSION TITLE
C# Markup: How to create templated Control?

### DISCUSSION THREAD

**Discussion by kucint** [unoplatform/uno#16284] [Q&A] [answered]
I am trying to create a simple templated Control using C# markup:

```
public sealed partial class MyControl : Control
{
    public MyControl() => this
        .Template(root => new Rectangle()
            .Fill(Colors.LightGray)
            .Stroke(Colors.Black)
            .StrokeThickness(1)
            .Width(100)
            .Height(100)
            );
}
```

This unfortunately crashes with following exception:

_Microsoft.UI.Xaml.Markup.XamlParseException: 'The text associated with this error code could not be found.
Failed to create a 'System.Type' from the text 'target:MyControl'. [Line: 6 Position: 3]'_

How am I suppose to do it correctly?

**Comment by dansiegel** [comment]
@kucint can you provide a simple reproduction of the issue. Also which platform are you hitting this bug on (or is it all of them)? Does it make any difference if you update to the latest 5.3 dev package?

**Comment by kucint** [comment]
The issue is observed on Windows (other platforms not tested).

```
{
  "msbuild-sdks": {
    "Uno.Sdk": "5.1.104",
    "Microsoft.Build.NoTargets": "3.7.56"
  }
}
```

The issue persist when upgraded to "Uno.Sdk": "5.3.0-dev.336":

![image](https://github.com/unoplatform/uno/assets/54866088/a864f802-5494-4769-9f45-9d38152fc171)

Here you have a simple reproduction of the issue:
[UnoApp16284.zip](https://github.com/unoplatform/uno/files/14995149/UnoApp16284.zip)

**Comment by kucint** [comment]
@dansiegel any news on it?

**Reply by dansiegel** [reply, marked as answer]
It should be fixed in 5.2 or the latest 5.3 dev packages. Please keep in mind that the fixes included some changes to the Generators so pre-generated dev packages other than Uno.WinUI.Markup may still have issues as they haven't yet been updated, but it should work just fine for the upcoming 5.2 release.

**Reply by kucint** [reply]
thanks for info!

**Reply by dansiegel** [reply]
No problem. Thanks for reporting the issue. Also just for clarity. The dev packages currently available for Themes, Toolkit, & Extensions just haven't been updated to be built with the latest version of the generator that includes part of the fix for this issue. Once we release 5.2 those repos will all get updated as well. The packages from those repos that are part of the 5.2 release cycle should be built against the 5.2 generator that has the fixes.

**Reply by kucint** [reply]
OK. Got it. Thanks!

FYI: I do observe other similar issues, like this one:
```
new Button()
 .Content("Click me!")
 .Command(p => { })  // this causes exception
```
throws _An exception of type 'System.InvalidOperationException' occurred in Uno.Extensions.Markup.dll but was not handled in user code_

I guess, all of them have the same origin as the current issue so I do not report them.... and just assume they all will be fixed in next release. Am I right?

**Reply by dansiegel** [reply]
This exception would actually make a ton of sense. If you look `p` is actually the `IDependencyPropertyBuilder` and since you didn't do anything with the IDependencyPropertyBuilder it will throw a runtime exception. C# Markup does not currently support inline Commands. You would need to bind to a command in your DataContext.

**Reply by francoistanguay** [reply]
Here's an example from our Counter App Tutorial: https://github.com/unoplatform/Uno.Samples/blob/6902af5627ae6b410fd5864231ffd1e2e6f9d7ca/reference/Counter/CSharp-MVVM/Counter/MainPage.cs#L33

**Comment by kucint** [comment]
> C# Markup does not currently support inline Commands. You would need to bind to a command in your DataContext.
Oh, so I must bind! This is new to me. This explain a lot. Thanks!
